packages/react-components/react-utilities/src/hooks/useMotionPresence.ts

```diff
@@ -0,0 +1,345 @@
+  has(property: string): boolean;
+  readonly size: number;
+}
+
+/**
+ * HTMLElement with styled map.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/computedStyleMap
+ */
+type HTMLElementWithStyledMap<TElement extends HTMLElement = HTMLElement> = TElement & {
+  computedStyleMap(): StylePropertyMapReadOnly;
+};
+
+interface CSSWithNumber {
+  number(value: number): {
+    value: number;
+    readonly unit: string;
+  };
+}
+
+/**
+ * State for useMotionPresence hook.
+ */
+export type UseMotionPresenceState<TElement extends HTMLElement> = {
+  /**
+   * Ref to the element.
+   */
+  ref: React.RefCallback<TElement>;
+
+  /**
+   * Whether the element should be rendered in the DOM.
+   * This should be used to conditionally render the element.
+   */
+  shouldRender: boolean;
+
+  /**
+   * Whether the element is currently visible in the DOM.
+   */
+  visible: boolean;
+
+  /**
+   * Current state of the element.
+   *
+   * - `entering` - The element is entering the DOM.
+   * - `exiting` - The element is exiting the DOM.
+   * - `resting` - The element is currently not animating. This is the final and initial state of the element.
+   * - `unmounted` - The element is not rendered in the DOM.
+   */
+  motionState: 'entering' | 'exiting' | 'resting' | 'unmounted';
+};
+
+/**
+ * Options for useMotionPresence hook.
+ */
+export type UseMotionPresenceOptions = {
+  /**
+   * Whether to animate the element on first mount.
+   *
+   * @default false
+   */
+  animateOnFirstMount?: boolean;
+};
+
+/**
+ * Returns CSS styles of the given node.
+ * @param node - DOM node.
+ * @returns - CSS styles.
+ */
+const getElementComputedStyle = (node: HTMLElement): CSSStyleDeclaration => {
+  const window = node.ownerDocument?.defaultView;
+
+  return window!.getComputedStyle(node, null);
+};
+
+/**
+ * Converts a CSS duration string to milliseconds.
+ *
+ * @param duration - CSS duration string
+ * @returns Duration in milliseconds
+ */
+function toMs(duration: string): number {
+  const trimmed = duration.trim();
+
+  if (trimmed.includes('auto')) {
+    return 0;
+  }
+
+  if (trimmed.includes('ms')) {
+    return parseFloat(trimmed);
+  }
+
+  return Number(trimmed.slice(0, -1).replace(',', '.')) * 1000;
+}
+
+/**
+ * Checks if the browser supports CSSOM.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/computedStyleMap
+ *
+ * @param node - DOM node
+ * @returns Whether the browser supports CSSOM
+ */
+const hasCSSOMSupport = (node: HTMLElementWithStyledMap) => {
+  /**
+   * As we are using the experimental CSSOM API, we need to check if the browser supports it.
+   * The typecast here is to allow the use of the `number` function that is not yet part of the CSSOM typings.
+   * @see https://www.npmjs.com/package/@types/w3c-css-typed-object-model-level-1
+   */
+  return Boolean(typeof CSS !== 'undefined' && (CSS as unknown as CSSWithNumber).number && node.computedStyleMap);
```

Comment:
```suggestion
  return Boolean(typeof CSS !== 'undefined' && (CSS as typeof CSS & CSSWithNumber).number && node.computedStyleMap);
```

Reply:
That's pretty clever, thanks. I changed the type definition so we can simplify the logic here.